### PR TITLE
Python Plugins

### PR DESCRIPTION
**Pull request by fpliger**
Goal of this PR is to add support for a very simple [and small] Python Plugins API so that PyScript users can implement Plugins in pure Python. The topic of Python Plugins opens up for a lot of questions and corner case scenarios, so the intention here is to add a small subset of the final features we envision to then iteratively receive feedback, learn from it, amend and reiterate. 

More explicitly, the goal of this work is to:

* add support to plugins that hook into the PyScript execution flow **after the interpreter has been initialized**  (i.e., `afterSetup`, `afterLaunch` and `onUserError`). This means that Python Plugins will not be notified of events that happened before it ( i.e.,  `configure` and `beforeLaunch`).
* add support to plugins that create new custom components that can be written in pure Python
* start small (with the simplest implementation/interface possible)and add features/optimize later. (i.e., the API is very simple and CustomElements plugins get the parent tag injected in `__init__` through injection)

Things explicitly left out of this work:

* Support for lazy events notification (basically receiving events that happened before the Python interpreter was ready
* plugins dependency specification (if a plugin depends on specific python packages, at the moment, it'll not auto-install anything and it's up to the client application to specify also the plugin dependencies)
* design choices on how Python Plugins will work in web workers
* design choices on an API that completely decouples the Plugin from implementation details (i.e., the JS bridge provided by Pyodide or access to DOM elements)

Things needed in this PR that were not part of the goal:

* changed the execution of `pyscript.py` to not be executed in the the global namespace but as a module
* added `fetchPythonPlugins`, that basically loads the plugin files into the FS and import them as module

### CODE REVIEW

File: pyscriptjs/src/plugin.ts [hunk 1]
```diff
@@ -57,19 +61,30 @@ export class Plugin {
...
+        this._pythonPlugins.push(plugin);
+    }
+
     configure(config: AppConfig) {
         for (const p of this._plugins)
             p.configure(config);
+
+        for (const p of this._pythonPlugins)
+            if (typeof p.configure !== 'undefined') {
+                p.configure(config);
```
Comment (ntoll) [inline pyscriptjs/src/plugin.ts:86]: You can do (in JavaScript, so I think it'll work in TypeScript):

```JS
p.configure?.(config);
```

Which will only be called if the `p.configure` method exists (saves the conditional).

**Review by ntoll — APPROVED** [review]
This all makes sense to me. Let's see where this gets us then refine, revise and refactor.

File: pyscriptjs/src/plugin.ts [hunk 2]
```diff
@@ -57,19 +61,30 @@ export class Plugin {
 
 export class PluginManager {
     _plugins: Plugin[];
+    _pythonPlugins: any[];
```
Comment (FabioRosado) [inline pyscriptjs/src/plugin.ts:64]: Since python plugins have the same lifecycle, we should perhaps create a type/interface for these and use it as type here instead of any

Comment (fpliger) [reply pyscriptjs/src/plugin.ts:64]: Yeah, we could. I didn't want to spend too much time on typing right now because, effectively, the objects being added here are actually PyProxy objects. This may change later though. We'll see

**Review by antocuni — CHANGES_REQUESTED** [review]
I have left many comments, but most are suggestions/discussions which you can decide to ignore.

IMHO the two most important points to solve are:
1. naming
2. the usage of the global `_pyscript_app`

It also lacks a lot of tests. It's a bit scary that you wrote this giant amount of code with only one test, I bet there are tons of lines which are just uncovered. I know that you plan to add more, but personally I think it's hard/impossible to arrive at a good coverage "after the fact".
I have the feeling that you will disagree with me here, but it's worth trying planting the seed :)

File: examples/markdown-plugin.html [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    <script defer src="../build/pyscript.js"></script>
+
+    <py-config>
+      packages = [
+        "markdown"
+      ]
+      plugins = [
+        "../build/plugins/python/py_markdown.py"
+      ]
+    </py-config>
```
[the remark this answers is not in the document]
Comment (JeffersGlass) [reply examples/markdown-plugin.html:20]: I was literally just typing the same thing! At least with regards to the naming. Was also thinking about the possibility of `js-plugins=[...]` at some point, which would presumably need to be treatly slightly differently in their loading.

File: pyscriptjs/src/plugin.ts [hunk 4]
```diff
@@ -85,10 +100,47 @@ export class PluginManager {
...
+            super();
+
+            this.shadow = this.attachShadow({ mode: 'open' });
+            this.wrapper = document.createElement('slot');
+            this.shadow.appendChild(this.wrapper);
+            this.source = this.innerHTML;
+            this.pyObject = pyObject(this);
+            if (typeof callback !== 'undefined') {
+                callback(this);
+            }
```
[the remark this answers is not in the document]
Comment (hoodmane) [reply pyscriptjs/src/plugin.ts:135]: `x !== undefined` raises a `ReferenceError` if `x` has not been declared, whereas `typeof x !== 'undefined'` allows for the case when no variable named `x` has been declared. 

In this case, `callback` is definitely defined since it's an argument. Since it should either be a function or undefined, I'd probably just go with `if(callback)`.

File: pyscriptjs/src/main.ts [hunk 5]
```diff
@@ -204,9 +207,34 @@ export class PyScriptApp {
...
+        runtime.globals.set('_pyscript_app', this);
+
+        // inject it into the PyScript module scope
+        // TODO: Currently adding the imports for backwards compatibility, we should
+        //       probably remove it
+        await runtime.run(`
+        import pyscript
+        pyscript.create_custom_element = create_custom_element
+        pyscript._pyscript_app = _pyscript_app
+        pyscript.PyScript.set_version_info('${version}')
```
Comment (hoodmane) [inline pyscriptjs/src/main.ts:227]: You could assign these attributes onto the Pyscript module without mucking with the Python global scope at all:
```js
let pyscript = runtime.pyimport("pyscript");
pyscript.create_custom_element = create_custom_element
pyscript._pyscript_app = this;
pyscript.PyScript.set_version_info(version);
pyscript.destroy();
```
generally I try to avoid string interpolating into code I'm going to exec when possible.

File: examples/markdown-plugin.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (fpliger) [reply examples/markdown-plugin.html:20]: I thought about it but couldn't find a "good" reason why we should create different keys to define the plugins I want to install in my app. Why do we need to add this complexity from a user standpoint?

There's more... imagine a future with a ton of plugins and a plugin index, it may not be clear if a plugin is implemented in `Python`, `JS` or..... something else. I'd not want to confuse users on that and would rather add a process where we are able to know what type of plugin is it that we are downloading (adding the concept of plugin manifest for instance).

File: pyscriptjs/src/main.ts [hunk 6]
```diff
@@ -204,9 +207,34 @@ export class PyScriptApp {
         // XXX: maybe the following calls could be parallelized, instead of
         // await()ing immediately. For now I'm using await to be 100%
         // compatible with the old behavior.
+        logger.info('importing pyscript');
+
+        // Save and load pyscript.py from FS
+        runtime.interpreter.FS.writeFile("pyscript.py", pyscript, { encoding: "utf8" });
```
[the remark this answers is not in the document]
Comment (fpliger) [reply pyscriptjs/src/main.ts:213]: Not really! Very weird tbh

File: pyscriptjs/src/main.ts [hunk 7]
```diff
@@ -204,9 +207,34 @@ export class PyScriptApp {
...
+
+        // inject it into the PyScript module scope
+        // TODO: Currently adding the imports for backwards compatibility, we should
+        //       probably remove it
+        await runtime.run(`
+        import pyscript
+        pyscript.create_custom_element = create_custom_element
+        pyscript._pyscript_app = _pyscript_app
+        pyscript.PyScript.set_version_info('${version}')
+        from pyscript import micropip, Element, console, document
```
[the remark this answers is not in the document]
Comment (fpliger) [reply pyscriptjs/src/main.ts:228]: That's why I put the comment above... The intention here is to basically preserve what's in the current release, add a deprecation (cycle) warning and then remove it on the following release. 

On the matter of what should be in the global namespace I think we agreed already that we'll start only with `display`. We'll disagree on things to be added in the future, for sure, but I thought we had agreement for now.

About `display`... that's an EXCELLENT question. :)

File: examples/markdown-plugin.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (antocuni) [reply examples/markdown-plugin.html:20]: +1 for having both JS and Python plugins in `config.plugins` *eventually*.

But my doubts are not about JS vs Python, they are about to to interpret the strings in a forward-compatible way. For example, consider this future use cases:
1. `plugins = ["./py_markdown.py"]`: this is your current semantics: fetch a single file from an URL
2. `plugins = ["https://..../py_markdown.py"]`: this works out of the box
3. `plugins = ["py_markdown"]`: fetch the `py-markdown` plugin from the POPI (PyScript Official Plugin Index 😅)
4. `plugins = ["foo", "bar.baz"]`: fetch the plugins `foo` and `bar.baz` from the POPI
5. `plugins = ["foo.py"]`: *ambigous*: is it a relative URL or a plugin called `foo.py` from POPI?

They all seem reasonable, but suddenly point (5) is ambiguous: sure, we could have heuristics to say that if something ends with `.py` is always an URL, or the opposite that something is always a POPI name unless you specify `https://` or `./`, but they look a bit fragile.
But most importantly, we *don't know yet* how we want `config.plugins` to look like in the future.

Basically, I'm just worried that if we use `config.plugins` *now*, it will cause headaches in the future. That's why I wanted to use "something else", and possibly something which makes it clear that it's a preview/alpha/experimental/whatever. So, few suggestions:
```
py_plugins = [...]
_plugins = [...]
experimental_plugins = [...]
plugins_but_dont_use_me_please = [...]   # :)

[experimental]
plugins = [...]

[unstable]
plugins = [...]
```

File: pyscriptjs/src/main.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (antocuni) [reply pyscriptjs/src/main.ts:213]: I think it's related to the other problem we discussed: I start to think that maybe `FS.writeFile` is not synchronous and must be awaited somehow.

Maybe @rth or @hoodmane knows?
For context, `runtime.interpreter` is the result of `loadPyodide()`: it seems that the files written in this way are not immediately available for `import` (i.e., sometimes you get `ModuleNotFoundError`, sometimes not). Does it ring any bell?

File: pyscriptjs/src/main.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (antocuni) [reply pyscriptjs/src/main.ts:228]: yes, that why I said +0. I just wanted to warn you about the future disagreements 😂

> About display... that's an EXCELLENT question. :)

I think we should investigate: it might be a hint that something very weird is going on

File: pyscriptjs/src/main.ts [hunk 8]
```diff
@@ -234,6 +262,37 @@ export class PyScriptApp {
...
+                const pathArr = singleFile.split('/');
+                const filename = pathArr.pop();
+                // TODO: Would be probably be better to store plugins somewhere like /plugins/python/
+                //       or similar
+                const destPath = `./${filename}`
+                await runtime.loadFromFile(destPath, singleFile);
+                const modulename = singleFile.replace(/^.*[\\/]/, '').replace('.py', '');
+
+                console.log(`importing ${modulename}`);
+                await runtime.run(`import ${modulename}`);
```
[the remark this answers is not in the document]
Comment (JeffersGlass) [reply pyscriptjs/src/main.ts:286]: We could add the [`await-thenable`](https://typescript-eslint.io/rules/await-thenable/) rule to our eslint config, which highlights any time you try `await` a non-thenable object. 

To make that work, we'd need a return type for `runtime.run` - I went down a little bit of a rabbit hole trying to figure out how to create a type for "Any Object that Not a Promise" in typescript. As far as I can tell, there's not a built-in way to do it, but I am fairly novice in TS.

One can do: 
```ts
type NotAPromise<T> = T & { then?: void};

abstract run(code: string): NotAPromise<unknown>;
   ...
```
Which works, and correctly catches the two uses of `await runtime.run(...)` in this branch... but I'm not certain how many crimes that Type is committing.

File: pyscriptjs/src/main.ts [hunk 9]
```diff
@@ -234,6 +262,37 @@ export class PyScriptApp {
...
+
+            } catch (e) {
+                //Should we still export full error contents to console?
+                handleFetchError(<Error>e, singleFile);
+            }
+        }
+        logger.info("All plugins fetched");
+    }
+
```
[the remark this answers is not in the document]
Comment (antocuni) [reply pyscriptjs/src/main.ts:295]: or maybe we could avoid/solve the code duplication by having some common function `fetchListOfURLs` which is called by both `fetch.ts` and `plugin.ts`

File: pyscriptjs/src/main.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (fpliger) [reply pyscriptjs/src/main.ts:213]: Yeah, correct. I think that's probably the reason.

[the remark this answers is not in the document]
Comment (hoodmane) [reply pyscriptjs/src/main.ts:213]: `FS.writeFile` is synchronous. The problem is that you need to call `importlib.invalidate_caches`.

File: pyscriptjs/src/main.ts [hunk 10]
```diff
@@ -234,6 +262,37 @@ export class PyScriptApp {
...
+        logger.info("Python plugins to fetch: ", plugins)
+        for (const singleFile of plugins) {
+            logger.info(`  fetching plugins: ${singleFile}`);
+            try {
+                const pathArr = singleFile.split('/');
+                const filename = pathArr.pop();
+                // TODO: Would be probably be better to store plugins somewhere like /plugins/python/
+                //       or similar
+                const destPath = `./${filename}`
+                await runtime.loadFromFile(destPath, singleFile);
```
[the remark this answers is not in the document]
Comment (fpliger) [reply pyscriptjs/src/main.ts:280]: Yes. Agree on both things... I'm on a rush right now so I won't change the sequential -> parallel aspected but it can be done as a follow up PR. In case no one takes care of it, I can do it when back after Thanksgiving

File: pyscriptjs/src/main.ts [hunk 11]
```diff
@@ -234,6 +262,37 @@ export class PyScriptApp {
...
+                const modulename = singleFile.replace(/^.*[\\/]/, '').replace('.py', '');
+
+                console.log(`importing ${modulename}`);
+                await runtime.run(`import ${modulename}`);
+
+            } catch (e) {
+                //Should we still export full error contents to console?
+                handleFetchError(<Error>e, singleFile);
```
[the remark this answers is not in the document]
Comment (fpliger) [reply pyscriptjs/src/main.ts:298]: Honestly, I didn't put too much thought and just copied the code from `fetchPaths` assuming it was "just fine". 

I'd say we open an issue for both at this point

File: pyscriptjs/src/main.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (fpliger) [reply pyscriptjs/src/main.ts]: I'm very -1 on this one. I think this would be a very bad UX from a plugins consumer standpoint. As a consumer I wouldn't want to care where pyscript put the plugins... I just want to specify where to load (~ fetch) them from.

File: pyscriptjs/src/main.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (hoodmane) [reply pyscriptjs/src/main.ts:213]: At least I hope that's the problem...

[the remark this answers is not in the document]
Comment (antocuni) [reply pyscriptjs/src/main.ts:213]: why do you need to call `invalidate_caches`? If we never tried an `import pyscript` before, it shouldn't matter, should it?

File: pyscriptjs/src/python/pyscript.py [hunk 12]
```diff
@@ -463,4 +464,23 @@ def uses_top_level_await(source: str) -> bool:
     return TopLevelAsyncFinder().is_source_top_level_await(source)
 
 
+class Plugin:
+    def __init__(self, name):
+        self.name = name
+        self.app = _pyscript_app
+        self.app.plugins.addPythonPlugin(create_proxy(self))
```
[the remark this answers is not in the document]
Comment (antocuni) [reply pyscriptjs/src/python/pyscript.py:474]: > honestly I don't really like the fact that we are looking at module.plugin to find the plugin to be imported

yes, let's discuss this separately, but just to start planting the seed, note that it's a standard pattern. For example, to load a WSGI application into `uwsgi`, you pass the option `--wsgi-file`: the webserver expects the module to [have a callable called `application`](https://uwsgi-docs.readthedocs.io/en/latest/WSGIquickstart.html#the-first-wsgi-application) (there are ways to customize the name it if you really want).

File: pyscriptjs/src/plugin.ts [hunk 13]
```diff
@@ -85,10 +100,47 @@ export class PluginManager {
...
 }
+
+
+export function create_custom_element(name: string, pyObject: any, callback?: Function) : any {
+    logger.info(`creating plugin: ${name}`)
+    class CustomElementWrapper extends HTMLElement {
+        shadow: ShadowRoot;
+        wrapper: HTMLElement;
+        pyObject: any;
+        source: string;
```
[the remark this answers is not in the document]
Comment (fpliger) [reply pyscriptjs/src/plugin.ts:125]: Yeah... it is hard :) So, I've changed:

* `create_custom_element` to `define_custom_element`
* `pyObject` to `pyPluginClass` and `this.pyObject` to `this.pyPluginInstance` (even if, thinking a bit more, these don't **need** to be a Python class/instance, can be anything really)
* `CustomElementWrapper` to `ProxyCustomElement` (even if I'm still not that happy...). I didn't think `CustomElement` or `MyCustomElement` were great because it's not conveying the fact that it's really an Proxy Custom Element that delegates logic to `this.pyPluginInstance`. 
* `source` to `originalInnerHTML`.  `innerHTML` isn't great because it actually clashes with the `HTMLElement.innerHTML` attribute and it'd also give the idea that it's the current `innerHTML` while it's really meant to be the original `innerHTML` (the reason it was named was because I started with the `py-script` tag and that coincided with the source code within the tag but that's not true for all Custom Elements anymore)

File: pyscriptjs/src/python/pyscript.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (fpliger) [reply pyscriptjs/src/python/pyscript.py:474]: Yeah. Let's get it working and discuss that after the dust settles

File: pyscriptjs/src/python/pyscript.py [hunk 14]
```diff
@@ -463,4 +464,23 @@ def uses_top_level_await(source: str) -> bool:
...
+
+    def register_custom_element(self, tag):
+        console.log(f"Registering {tag}")
+
+        def wrapper(class_):
+            console.log(f"Inside dec {tag}")
```
[the remark this answers is not in the document]
Comment (fpliger) [reply pyscriptjs/src/python/pyscript.py:483]: these are for debugging :)

**Comment by fpliger** [comment]
Ok, I think it's ready for review. I still had a few things I wanted to fix/change but I think it's good enough to be merged as a first pass to go in the release and then fix in follow-up PRs. 

The biggest public facing thing I didn't have time to do is to actually move plugins under an `[experimental]` section. If we want to do that, @FabioRosado do you want to take it on a follow-up PR before we release?

Other things I think are missing but I think we can live with:

* There are still a couple of discussion topics to be discussed and then change later
* Need more tests
* I think I missed a docstring or 2

File: examples/markdown-plugin.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (antocuni) [reply examples/markdown-plugin.html:20]: +1 for `experimental.plugins`, so that we are free to completely break it when we move to the main section

File: pyscriptjs/src/main.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (antocuni) [reply pyscriptjs/src/main.ts]: +1 for having something which catches it.
I don't have any preference w.r.t. how to achieve that magic in typescript 😅

File: pyscriptjs/tests/integration/test_plugins.py [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+"""
+
+
+class TestPlugin(PyScriptTest):
+    def test_py_plugin_inline(self):
+        self.writefile("py_upper.py", PLUGIN_CODE)
+
+        # NOTE: sleep is needed here to avoid flakiness between the file above being written
+        #       and sometimes not yet "available"
+        sleep(1.5)
```
[the remark this answers is not in the document]
Comment (antocuni) [reply pyscriptjs/tests/integration/test_plugins.py]: well no, we figured out that the flakyness is inside the JS code and this sleep doesn't help at all

File: pyscriptjs/src/python/pyscript.py [hunk 16]
```diff
@@ -463,4 +464,27 @@ def uses_top_level_await(source: str) -> bool:
...
+    def register_custom_element(self, tag):
+        # TODO: Ideally would be better to use the logger.
+        console.info(f"Defining new custom element {tag}")
+
+        def wrapper(class_):
+            # TODO: this is very pyodide specific but will have to do
+            #       until we have JS interface that works across interpreters
+            define_custom_element(tag, create_proxy(class_))  # noqa: F821
+
+        return create_proxy(wrapper)
```
Comment (antocuni) [inline pyscriptjs/src/python/pyscript.py:487]: this should be `def define_custom_element`, should'nt it?

File: pyscriptjs/src/main.ts [hunk 17]
```diff
@@ -204,9 +201,34 @@ export class PyScriptApp {
...
+        //       remove it
+        await runtime.run(`
+        from pyscript import *
+        `);
+        logger.warn(`DEPRECATION WARNING: 'micropip', 'Element', 'console', 'document' and several other \
+objects form the pyscript module (with the exception of 'display') will be \
+be removed from the Python global namespace in the following release. \
+To avoid errors in future releases use import from pyscript instead. For instance: \
+from pyscript import micropip, Element, console, document`);
+
```
Comment (antocuni) [inline pyscriptjs/src/main.ts:225]: -1 for this: it's a warning which is easy to overlook and most importantly it is shown unconditionally, which will probably confuse people. I'd prefer something along these lines:
```python
class DeprecatedGlobal:
    def __init__(self, name, obj):
        self.name = name
        self.obj = obj
        self.warning_already_shown = False

    def show_warning_maybe(self):
        if self.warning_already_shown:
            return
        showWarning(f"Direct usage of {self.name} is deprecated. Please use pyscript.{self.name} instead")
        self.warning_already_shown = True

    def __getattr__(self, attr):
        self.show_warning_maybe()
        return getattr(self.obj, attr)

    def __call__(self, *args, **kwargs):
        self.show_warning_maybe()
        return self.obj(*args, **kwargs)

Element = DeprecatedGlobal('Element', pyscript.Element)
console = DeprecatedGlobal('console', pyscript.console)
...
```

But I think it's fine to do this in a follow-up request (I'm happy to do it if you want)

File: pyscriptjs/src/plugin.ts [hunk 18]
```diff
@@ -1,9 +1,11 @@
 import type { AppConfig } from './pyconfig';
 import type { Runtime } from './runtime';
 import type { UserError } from './exceptions';
+import { getLogger } from './logger';
 
-export class Plugin {
+const logger = getLogger('pyscript/main');
```
Comment (antocuni) [inline pyscriptjs/src/plugin.ts:6]:
```suggestion
const logger = getLogger('plugin');
```

File: pyscriptjs/src/plugins/python/py_markdown.py [hunk 19]
```diff
@@ -0,0 +1,31 @@
+from textwrap import dedent
+
+from markdown import markdown
+from pyscript import Plugin, console
+
+console.warning(
+    "WARNING: This plugin is still in a very experimental phase and will likely change"
+    " and potentially break in the future releases. Use it with caution."
+)
+
```
Comment (antocuni) [inline pyscriptjs/src/plugins/python/py_markdown.py:10]: again -1 on this. If it's a warning that the user should see, we should display it on the DOM (and add a way to silence it in the config)

**Review by antocuni — APPROVED** [review]
The diff *huge*. It seems that you have applied some kind of autoformatter which modified code in random files.
Ideally, I would like to have a clean diff for the sake of history, but I have the feeling that this is too much to ask 😅.

Anyway, this probably means that we should add a JS formatter in the pre-commit config, to ensure consistency and make sure that this won't happen in the future

File: pyscriptjs/src/main.ts (hunk 17 above)
Comment (antocuni) [reply pyscriptjs/src/main.ts:225]: In the meantime, I found out that the mess is even worse that this, see  https://github.com/pyscript/pyscript/issues/986

Comment (fpliger) [reply pyscriptjs/src/main.ts:225]: Meh... I'm -0 because I don't think it's worth the effort and it's not user facing (when user == users of the PyScript application, not devs).

If you want to take a stab at it and wouldn't delay the release even more, I'm not opposed to that

File: pyscriptjs/src/plugins/python/py_markdown.py (hunk 19 above)
Comment (fpliger) [reply pyscriptjs/src/plugins/python/py_markdown.py:10]: On this one, we are in strong disagreement here. I'm -100 on raising any warning that is so invasive that we put on the DOM. As a developer user of PyScript I'd be really really frustrated that my users have to see a weird warning that they don't understand and adds -10 value to their experience on my app.

**Comment by JeffersGlass** [comment]
Hi Y'all - minimal time today contribute today, my apologies, but did want to chime in on one of the failing tests, since some of the reason they failed is because of changes yesterday:

- `FAILED test_runtime_config` - this is failing because this test loads pyodide v20, and pyscript.py now tries to import `pyodide.ffi`, which was only introduced in v21. 

Perhaps use v21.0 as a test version? (PyScript currently on v21.3)

**Comment by FabioRosado** [comment]
> Hi Y'all - minimal time today contribute today, my apologies, but did want to chime in on one of the failing tests, since some of the reason they failed is because of changes yesterday:
> 
> * `FAILED test_runtime_config` - this is failing because this test loads pyodide v20, and pyscript.py now tries to import `pyodide.ffi`, which was only introduced in v21.
> 
> Perhaps use v21.0 as a test version? (PyScript currently on v21.3)

Yeah I was looking at CI and there are a few other things failing, I'm going to take a look in a second and fix conflicts :)

**Comment by fpliger** [comment]
@JeffersGlass thanks for the hint on the pyodide API change. TS tests are fixed locally, hope it does the trick

**Comment by FabioRosado** [comment]
It seems that importing modules is broken in CI, but so far I haven't been able to figure out why 🤔

File: pyscriptjs/src/plugins/python/py_markdown.py (hunk 19 above)
Comment (antocuni) [reply pyscriptjs/src/plugins/python/py_markdown.py:10]: our users are the developers, and by having "hidden warnings" we don't do a good service to them.
*Their* users are the final users of the app and yes, the final users don't want to see a warning: that's why there should be an option to silence it, so that *our* user can decide whether to fix the problem or hide the warning.

File: pyscriptjs/src/main.ts (hunk 17 above)
Comment (antocuni) [reply pyscriptjs/src/main.ts:225]: > Meh... I'm -0 because I don't think it's worth the effort and it's not user facing (when user == users of the PyScript application, not devs).

I think that this is the core of many disagreements. From my point of view, our users are the developers, and we should try to make the life simpler to them.
It is *their* job to make nice apps for their final users, not ours; our job is to make it as simple as possible to make it so.

> If you want to take a stab at it and wouldn't delay the release even more, I'm not opposed to that

doing a release even if there is a known broken feature "because it's late" doesn't sound like a good idea. We will do a release when it's ready, IMHO.
The history tells us that putting features in a rush "because we need to release" doesn't end well ;)

That said, I'm happy to do it.
